### ISSUE TITLE
GraphQL WordPress Plugin Performance Issue

### ISSUE THREAD

**Issue by gagan0123** [wp-graphql/wp-graphql#1873]
GraphQL Plugin's performance deteriorates significantly with the number of post-types and taxonomies defined in the WordPress.

### Reproduce the bug:
1. Try to add more than 20-30 post types, with around the same number of taxonomies defined for them. 
2. Check out the response timing of GraphQL endpoint.

To further add to this issue, even caching does not resolve this performance problem. I've tried with 
1. Object Caching (Redis)
2. MySQL caching(I know I shouldn't have)
3. Using a [GraphQL caching plugin for WordPress](https://github.com/valu-digital/wp-graphql-cache)

Screencast for the same:

[![](http://img.youtube.com/vi/Et7iFbV-H8A/0.jpg)](http://www.youtube.com/watch?v=Et7iFbV-H8A "GraphQL WP Plugin Performance Issue")

### Query statistics:
| Request Type:                                | WordPress Homepage | GraphQL Request(just endpoint) | WP-JSON Request(just endpoint) | GraphQL Request(Single Post) | WP-JSON Request(Single Post) |
|----------------------------------------------|--------------------|--------------------------------|--------------------------------|------------------------------|------------------------------|
| Timing with default WordPress                | 85 ms              | 220 ms                         | 80 ms                          | 230 ms                       | 80 ms                        |
| Timinig with 10 post types and 10 taxonomies | 90 ms              | 400 ms                         | 110 ms                         | 407 ms                       | 100 ms                       |
| Timing with 50 post types and 50 taxonomies  | 160 ms             | 1650 ms                        | 260 ms                         | 1650 ms                      | 230 ms                       |

Here's a link to XDebug profiling for the above requests(I've skipped the 10 post type in there)
https://drive.google.com/drive/folders/1-U_8Gb4BsL_UEHMyO0NKthRRkXQmq_2i?usp=sharing

#### What version of `WPGraphQL` you're using, and the platform(s) you're running it on
`WPGraphQL` version: 1.3.5

#### What other plugins you're using
No plugin active on the site except `WPGraphQL`

#### The behavior you expect to see, and the actual behavior
The number of post types and taxonomies defined in a WordPress site should have minimal/no impact on the performance of GraphQL queries. 

Any website running WooCommerce, has multiple post types defined for things like:
1. Orders
2. Products
3. Variations
4. Coupons

having custom taxonomies like:
1. Product type
1. Product visibility
1. Product categories
1. Product tags
1. Product shipping classes
1. Product Color
1. Product Design
1. Product Material
1. Product Size
1. Product Style

Once you set this up, and add a few more plugins that adds CPTs and site almost becomes unusable, if built using GraphQL.

### When reporting a _bug_, please be sure to include the following:
- [x] A descriptive title
- [x] An *isolated* way to reproduce the behavior (example: GitHub repository with code isolated to the issue that anyone can clone to observe the problem)
- [x] What version of `WPGraphQL` you're using, and the platform(s) you're running it on
- [x] What other plugins you're using
- [x] The behavior you expect to see, and the actual behavior

**Comment by andreasradloff** [comment]
I noticed this too. Schema generation seems a bit slow and seems to occur on every request. Could it be cached somehow?

**Comment by PaperPlaneSoftware** [comment]
I've also noticed that WPGraphQL takes a rather long time to respond.

<img width="960" alt="Screenshot 2021-07-06 114507" src="https://user-images.githubusercontent.com/25370730/124595198-a1532080-de58-11eb-8197-22446d6b5109.png">

Making multiple graphql queries which take 1 - 3s to respond makes the site very slow.

**Comment by gagan0123** [comment]
In the XDebug profiler data, we can clearly see that the `get_schema` method of `SchemaRegistry` is what overwhelming the request.
![image](https://user-images.githubusercontent.com/1535505/124596444-e8084100-de7f-11eb-9146-519a11b45c3b.png)
If we can even cache that out, maybe in transient or something, this could significantly improve the performance.

**Comment by andreasradloff** [comment]
My first thought was a transient so I took a stab at that but since the schema is a complex object it is not serializable in a way that transients can cache. Some new code would be needed to create a cacheable object (like an array) from the schema and a way to reconstruct the schema using that object.

**Comment by andreasradloff** [comment]
Could we get some input from the maintainers, @jasonbahl maybe, if this is has been addressed before or any ongoing work on improving performance of schema generation?

**Comment by andreasradloff** [comment]
Sorry for the comment spam but it might be worth investigating if the Symfony framework serializer class might be a good fit for this. https://symfony.com/doc/6.0/components/serializer.html

**Comment by andreasradloff** [comment]
I made an attempt to chace using the symfony serializer but the process is heavy and runs out of memory, maybe some circular dependencies that can be avoided. Here's the code in case someone feels like continuing the work. 

`composer require symfony/serializer`


```php

use Symfony\Component\Serializer\Encoder\XmlEncoder;
use Symfony\Component\Serializer\Normalizer\ObjectNormalizer;
use Symfony\Component\Serializer\Normalizer\PropertyNormalizer;
use Symfony\Component\Serializer\Normalizer\AbstractNormalizer;
use Symfony\Component\Serializer\Serializer;

  public static function get_schema() {
    $cache = get_transient( 'wpGraphQLSchema');
    if ($cache) return self::$schema = $cache;
    if ( null === self::$schema ) {
      $defaultContext = [
          AbstractNormalizer::CIRCULAR_REFERENCE_HANDLER => function ($object, $format, $context) {
            return isset($object->name) ? $object->name : $object::class;
          },
      ];
      $normalizer = new PropertyNormalizer(null, null, null, null, null,  $defaultContext);
      $encoders = [new XmlEncoder()];
      $normalizers = [$normalizer, new ObjectNormalizer()];
      $serializer = new Serializer($normalizers, $encoders);
      $cache = get_transient( 'wpGraphQLSchema');
      if ($cache) {
        self::$schema = $serializer->deserialize($cache, WPSchema::class, 'xml');
      } else {
        $schema_registry = new SchemaRegistry();
        $schema          = $schema_registry->get_schema();
        /**
         * Generate & Filter the schema.
         *
         * @since 0.0.5
         *
         * @param array                 $schema      The executable Schema that GraphQL executes against
         * @param AppContext $app_context Object The AppContext object containing all of the
         *                                           information about the context we know at this point
         */
        self::$schema = apply_filters( 'graphql_schema', $schema, self::get_app_context() );
        try {
          $cache = $serializer->serialize(self::$schema, 'xml');
          set_transient( 'wpGraphQLSchema', $cache, 100);
        } catch (Exception $exception) {
          error_log('WPGraphQL: ' . $exception->getMessage());
        }
      }
}```

**Comment by chriszarate** [comment]
Just noting that I'm following along and will try to take a look to see if I can help. Thanks for the report!

**Comment by jasonbahl** [comment]
I believe the culprit here is that we're loading ALL types for each GraphQL request, and not _just_ the Types that are needed to fulfill the request. 

Ideally, a query such as: 

```graphql
{
  posts {
    nodes {
      id
    }
  }
}
```

Should only generate the following Types: 

- RootQuery
- RootQueryToPostConnection
- Post

As these are the Types that are needed to validate and execute the actual query. 

I believe what's happening (as seems clear by what's presented here) is that we're loading the full schema and ALL types on every request, instead of lazy loading the types. 

I think there's some work we can do to delay registration of Types and load them only when needed instead of load them on every request. 

I'm not sure how "easy" this will be to accomplish, but something @markkelnar and I are looking into and glad ya'll are interested in looking into it as well! 

@markkelnar and I are also working on other performance improvements including Persisted Queries and other caching mechanisms.

**Comment by chriszarate** [comment]
It seems like WPGraphQL is using `typeLoader` as specified here:

https://webonyx.github.io/graphql-php/schema-definition/#lazy-loading-of-types

Perhaps this isn't working as intended?

**Comment by jasonbahl** [comment]
@chriszarate yes, I think that's the case. 

I think what's happening is that when we register a type, it's building the object at registration time, but what should happen is that it registers a callback, and the callback is executed during lazy loading and so the Type would only be generated when it's needed instead of up front. 

I _think_ there's a path forward where we can refactor a few functions in TypeRegistry.php to fix this 🤞🏻

**Comment by drtechie** [comment]
Hello,

First of all thank you for this great plugin!
We have been using this plugin since 0.3.x.

**Stack and incident details**
Of late we have encountered some nasty performance issues. We use a single instance of WordPress to support content across multiple implementations. We have around 6 post types, 125 tags, 40 categories and each of these post types have anywhere between 30 - 50 custom fields.

In addition to wpgraphql plugin, we also have WPGraphQL Meta Query, WPGraphQL Tax Query and WPGraphQL for Advanced Custom Fields plugins with the latest WordPress (5.7.2).

**Our findings**

In the last 2 weeks, we have onboarded a few thousand users to our app(s) (>50K users). Requests to `/graphql` showed a significant increase in response time. 

<img width="1105" alt="image" src="https://user-images.githubusercontent.com/1007084/125120576-3fbcdb80-e110-11eb-9d06-0ee92271fb63.png">


note the `rt=` here is the response time and it's over 30 seconds.

1. The response times start with 1-2 seconds and degrades over time when more and more users start using the app(s). 
2. What is interesting is that CPU and memory utilization was low but load average was consistently high (2.x-3.x) on a 2 core system. 
3. GraphQL tracing showed that the total execution period was very low.
<img width="460" alt="image" src="https://user-images.githubusercontent.com/1007084/125118051-c66fb980-e10c-11eb-97fd-0f63ded4da0b.png">

4. I did benchmark against a staging server (2 core, 1GB) and got 4.x requests/second response. Against Rest API, I got 16.x requests/second. (this means GraphQL API was running at least 4 times slower). Again the interesting thing is that the GraphQL plugin reports a fraction of the second as the original query time (0.13 seconds out of 2000ms total response time). 

![image](https://user-images.githubusercontent.com/1007084/125118256-13ec2680-e10d-11eb-9878-d1741ef4db05.png)

5. Following changes were done to our production system.
We scaled up the system to 2-core, 8GB one. The same symptom persisted. After restarting services, the response time starts with a few seconds and degrades quickly to 20 seconds.

6. The number of PHP-FPM processes was set to 3. On increasing to 10, the load average also increased to 7 / 8!

7. We changed the system further up to a 4 core, 16GB one. Right now the response times have finally calmed down and are in an acceptable range.
<img width="929" alt="image" src="https://user-images.githubusercontent.com/1007084/125118961-0d11e380-e10e-11eb-9567-023d8abce2e4.png">
The addition of CPU cores reduced the load significantly.

8. In addition to the above, I have tried various caching plugins but that did not help.

Let me know if you need any more metrics and information.
Looking forward to a fix - quite certain we will hit another bottleneck soon with our growth rate. :)

Thanks,

**Comment by jasonbahl** [comment]
Closed by #2000

**Comment by gagan0123** [comment]
Thanks @jasonbahl for fixing it up :)

I can definitely see the improvement in performance on the site I worked on.

**Comment by joaovpmamede** [comment]
Is this really fixed?
I'm seeing a lot of cpu being used by `php-fpm` everytime I make a call to `/graphql`.
This is happening a lot with a nextjs site not only during the build process (which tries to create html pages) but also when I fetch data from the client (happens with a simple query through Postman as well).

**Comment by jasonbahl** [comment]
@joaovpmamede what version of WPGraphQL are you on? 

> I'm seeing a lot of cpu being used by php-fpm everytime I make a call to /graphql.

Does this happen when you hit the `/graphql` endpoint with no query at all?  Or does it only happen when you are making queries for data. 

This particular fix (#2000) fixed an issue that affected _all_ requests, including requests that didn't even query data. 

If you try WPGraphQL 1.5.* or earlier and compare it with 1.6*+, I think you'll see the difference. 

Is there still queries that will take resources to resolve, absolutely! Can they be improved? most certainly! I think we need to narrow down more specifically the types of queries that are causing issues and troubleshoot them more specifically at this point. 

If you can share more information about what queries seem to lead to more CPU usage than others, that would be great. 

Also, if you can try things with/without other plugins/custom code in place, that would be good as well. 

The more information you can provide about your specific issues, the better we can identify underlying issues and come up with solutions.

**Comment by joaovpmamede** [comment]
@jasonbahl hitting `/graphql` seems to be fine and I don't see nothing happening on the CPU.
But a simple query as this one:

<img width="358" alt="Screenshot 2022-03-11 at 16 26 46" src="https://user-images.githubusercontent.com/1040722/157907937-7bdd47d1-9af0-4f7b-be9e-902952194b3e.png">

creates some load and takes almost 300ms.

Plugins I'm using with wp-graphql (v1.6.12, although I've tried the 1.7.1 too):

- ACF Pro
- Polylang
- The SEO Framework
- ACF Options for Polylang
- WP GraphQL Polylang
- WPGraphQL for Advanced Custom Fields
- WPGraphQL for The Seo Framework
- WPGraphQL JWT Authentication
- WPGraphQL Meta Query
- WPGraphQL Offset Pagination
- WPGraphQL Tax Query

I'll continue to investigate this and I'm trying to profile but without success.

**Comment by jasonbahl** [comment]
@joaovpmamede One thing I would recommend when using ACF is setting up their Local JSON feature (https://www.advancedcustomfields.com/resources/local-json/), if you haven't already. 

That should help reduce load significantly, both in context of WPGraphQL, but even the WordPress dashboard.

**Comment by jasonbahl** [comment]
It might be worth doing some simple things, like make the same query you shared with no plugins active. Is it faster than 300ms or the same? Then activate the plugins one at a time. See if one of the plugins is causing slowdowns, or if, perhaps, the server is generally slow.

**Comment by jasonbahl** [comment]
![CleanShot 2022-03-11 at 09 37 03](https://user-images.githubusercontent.com/1260765/157909259-87e98355-ba94-4f0f-a90b-7d44c739615d.png)

3.3s for the pages endpoint (which returns 10 pages).

I'm seeing the REST API responses aren't exactly fast on the same site (https://flowershopcolombiacheckout.com/wp-json/wp/v2/pages)

A GraphQL Query for 10 pages is faster than the REST endpoint (1.85s)

![CleanShot 2022-03-11 at 09 38 38](https://user-images.githubusercontent.com/1260765/157909503-4f48f26b-4359-48ea-9b69-670b6133d881.png)

**Comment by joaovpmamede** [comment]
@jasonbahl yup, we're using the Local JSON feature with ACF.

I'll try disabling the plugins although I've already made some tests and it doesn't really help that much - still see a lot o CPU usage.

Btw, that's not my website. More info [here](https://github.com/WebDevStudios/nextjs-wordpress-starter/discussions/968) as I've decided to test different hosts/stacks.

**Comment by jasonbahl** [comment]
> Btw, that's not my website

@joaovpmamede ah. sorry. Must've got that link confused with another issue. 🤷🏻‍♂️ 

In any case, might be good to see if your REST endpoint(s) are performing any different than your GraphQL endpoints. 

Often times, there can be an underlying issue that is impacting both. For example, the ACF Local JSON thing. It will make every use of ACF faster, the admin, GraphQL and REST endpoints.

**Comment by intelligence** [comment]
My suspicion is that ACF is causing performance issues. It would be nice to know if you discover the culprit @joaovpmamede.

**Comment by idflood** [comment]
I had similar slow requests as @joaovpmamede and tried to find what was the culprit. I still haven't found the cause but one thing I noticed is that the REST endpoint(s) are noticeably faster (300ms vs 1.4 second or more for simple graphql queries)

In the tests below I'm using a docker environment with PHP 8, latest version of Wordpress and plugins. I also quickly checked the same site on MAMP. The response times are a bit faster but it doesn't change the overall difference between rest and graphql.

We were also using a legacy "Full theme" with a lot of unneeded function and some composer dependencies so I also tested with a super minimal theme (only adding a few content types and taxonomies). There is in total **4 custom post types** and **5 custom taxonomies**, removing them didn't make any relevant difference. Also note that we are already using the _acf-json_ folder.

| Full theme + all plugins | ACF disabled | Full theme + no plugins | Basic theme + no plugins |
| --- | --- | --- | --- |
| 1.8s | 1.7s | 1.4s | 850ms |

For comparison a call to `/wp-json/wp/v2/pages/` takes less than **350ms**, even with the full theme and all plugins. The noderesolver pull request https://github.com/wp-graphql/wp-graphql/pull/2680 version didn't make any difference in response time.

I would love to have the same speed as rest endpoints for graphql. Let me know if there is something else I can try.

**Comment by justlevine** [comment]
> I would love to have the same speed as rest endpoints for graphql. Let me know if there is something else I can try.

@idflood could you perhaps run a full profile trace of both the graphql and rest requests and share the results? Would help pinpoint where exactly the slowdown is happening.

**Comment by idflood** [comment]
@justlevine good idea. Here is a first profiling, with on the left a request to wp-graphql and on the right to REST.

![Screenshot 2023-01-20 at 13 05 55](https://user-images.githubusercontent.com/197418/213691823-45bf8089-f687-45ea-8c58-7ef251316780.png)

Not sure if I interpret this correctly but I see a lot of time spent on is_readable, and if I compare the callers of that method it's called 4371 times from load_textdomain with wp-graphql and only once with REST.

![Screenshot 2023-01-20 at 13 14 00](https://user-images.githubusercontent.com/197418/213692988-95366f22-c925-426d-b6b1-7a14345cb993.png)

And for references the callers for load_textdomain. There might be something to optimize with our "full theme" but eventually it should not make such difference since the same is also enabled when calling rest endpoints.


![Screenshot 2023-01-20 at 13 23 16](https://user-images.githubusercontent.com/197418/213694390-67b760bd-3421-496f-a152-7dabadcb730b.png)

**Comment by idflood** [comment]
Maybe it's just `__` which is called a lot with wp-graphql, again with graphql on the left and rest on the right.

The issue might be more visible with a "slow" filesystem (like docker on osx)
We also usually install Wordpress in french, so it may load more translation files.


![Screenshot 2023-01-20 at 14 07 09](https://user-images.githubusercontent.com/197418/213702718-fd52d975-6453-438a-8ae7-e4f2e5efacc8.png)

**Comment by justlevine** [comment]
Was thinking the same thing. Pretty much every single graphql field has a semantic description, which looks like are getting translated on every single query even though they dont really need to be....
(Also puzzled by the huge increase in `get_option()` calls, but compared to `load_textdomain()` its negligible)

Either way at least gives us enough data to try to replicate (vanilla install in `en`, and then test again with a different language).

@jasonbahl should we reopen this ticket or move this over to a new `issue`?

**Comment by idflood** [comment]
@justlevine yes, here is a last screenshot sorted by called count, apply_filters also has a big difference.

I will push a basic project on github to reproduce the issue as soon as possible.

![Screenshot 2023-01-20 at 14 30 36](https://user-images.githubusercontent.com/197418/213707190-6bdf38ff-f620-4ded-bab8-9db4f0da5345.png)

**Comment by idflood** [comment]
@justlevine here is a minimal test repo https://github.com/idflood/wp-graphql-1837-perf
It's on that project where time go from 1.4s to 750ms by switching the site language. Note that I added a few plugins since having more translations files make the issue more visible.

**Comment by idflood** [comment]
@jasonbahl let me know if you want to reopen this issue or if I should create a new ticket.

**Comment by idflood** [comment]
I quickly created a [test branch](https://github.com/idflood/wp-graphql-1837-perf/tree/quicktest-no-descriptions-wp-graphql) on the minimal test repo with a modified version of wp-graphql.

In that version I simply commented nearly all the `description` fields in wp-graphql.

Response time for a simple query went from ~1.4s to ~950ms so this confirm that these `__()` calls contribute to these slow queries.

**Comment by creative-andrew** [comment]
@idflood  I am having exactly the same problem. https://wp-graphql.slack.com/archives/C3NM1M291/p1676113374565389

I went from 3 secs to 600ms just by switching to English from german. 

As a hack, I am considering disabling translations for the text domains of the WpGraphQL family of plugins. 

![image](https://user-images.githubusercontent.com/18492002/218270204-0989ed55-df8a-4dd7-9358-0dc47e688fe6.png)
![image](https://user-images.githubusercontent.com/18492002/218270208-a147ab67-2f0f-4677-8036-717e0c7fcee2.png)

**Comment by jasonbahl** [comment]
I see #2721 was opened. I think we should open isolated issues, and can reference this one for context as needed.

**Comment by jasonbahl** [comment]
@idflood what queries are you running to profile? 

If you query something like `{ posts { nodes { id, title } } }` are you seeing _that many_ calls to the `__()` translation function? Or is it only happening when full introspection queries are ran, like those which generate the Schema for tools such as the GraphiQL IDE?

I would expect there to be a LOT of calls to the `__()` function when loading the whole Schema, but not when executing specific queries.

**Comment by idflood** [comment]
@jasonbahl I ran your example query from postman, and with the site en french I got 4241 calls to `__()`. I've also double checked that the introspection, debug and logs option where turned off in wp-graphql settings.

Here is a screenshot of the profile I did this morning, keeping it as a link since it's similar to the previous ones.
https://user-images.githubusercontent.com/197418/218964647-a8bcb233-d9bc-43ca-99b5-c2214d7aed00.png

And here is the cachegrind profile file if you want to dig deeper.
https://www.swisstransfer.com/d/ef3fbf27-70b8-4234-a9f7-346bf42bd630
